### PR TITLE
ci: Add GitHub action for core build and python testing in pull requests

### PR DESCRIPTION
**Pull request by pranavm-nvidia**

### CODE REVIEW

File: python/test/test_api.py [hunk 1]
```diff
@@ -359,8 +359,9 @@ def test_stop(self):
                         "parameters": {"decoupled": {"string_value": "False"}},
                         # Keep instance count low for fast startup/cleanup.
                         # Alternatively can use KIND_CPU here, but keeping gpus/count explicit.
+                        # TODO: Check above
                         "instance_group": [
-                            {"kind": "KIND_GPU", "gpus": [0], "count": 1}
+                            {"kind": "KIND_CPU"}
```
Comment (pranavm-nvidia) [inline python/test/test_api.py:364]: This allows us to run this test on machines without GPUs, but the comment suggests it was intentionally run on GPU. If that's case, we will need to skip it in CI.

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        apt install -y --no-install-recommends clang-format-15 cmake libb64-dev rapidjson-dev libre2-dev
+        wget -O /tmp/boost.tar.gz https://archives.boost.io/release/1.80.0/source/boost_1_80_0.tar.gz && (cd /tmp && tar xzf boost.tar.gz) && mv /tmp/boost_1_80_0/boost /usr/include/boost
+        pip install build pytest
+
+    - name: Build
+      run: |
+        mkdir -p /core/build
+        cd /core/build
+        cmake -DCMAKE_INSTALL_PREFIX:PATH=`pwd`/install -DTRITON_CORE_HEADERS_ONLY=OFF ..
+        export TRITON_PYBIND="_c/triton_bindings.cpython-310-x86_64-linux-gnu.so"
```
Comment (pranavm-nvidia) [inline .github/workflows/build-and-test.yml:33]: Need to remove this once my packaging tests are in

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+name: Build And Test
+
+on:
+    pull_request:
+      branches:
+        - main
+      types: [synchronize, opened, reopened, ready_for_review]
```
Comment (rmccorm4) [inline .github/workflows/build-and-test.yml:7]: NOTE: Essentially to skip running action when PR is a `draft`

File: python/test/test_api.py (hunk 1 above)
Comment (rmccorm4) [reply python/test/test_api.py]: This test isn't doing anything GPU specific 👍

**Review by rmccorm4 — APPROVED** [review]

**Pull request merged**